B Tree index configuration [HZ-1272]

TomaszGaweda opened the pull request:
Add B-Tree tiered storage configuration:
- page size to be used
- memory tier configuration like capacity

Client protocol PR: https://github.com/hazelcast/hazelcast-client-protocol/pull/428

Checklist:
- [x] Labels (`Team:`, `Type:`, `Source:`, `Module:`) and Milestone set
- [x] Label `Add to Release Notes` or `Not Release Notes content` set
- [ ] Request reviewers if possible
- [x] Send backports/forwardports if fix needs to be applied to past/future releases
- [x] New public APIs have `@Nonnull/@Nullable` annotations
- [x] New public APIs have `@since` tags in Javadoc

hz-devops-test commented:
The job `Hazelcast-pr-builder` of your PR failed. (Hazelcast internal details: [build log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/12059/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/12059/Hazelcast-pr-builder-12059.zip)).
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
---------SUMMARY----------
--------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-checkstyle-plugin:3.1.2:checkstyle (default) on project hazelcast: An error has occurred in Checkstyle report generation. Failed during checkstyle execution: There are 3 errors reported by Checkstyle 8.38 with /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/checkstyle/checkstyle.xml ruleset. -> [Help 1]
--------------------------
---------ERRORS-----------
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/config/BTreeIndexConfig.java:130:36: '+' should be on a new line. [OperatorWrap]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/config/BTreeIndexConfig.java:131:40: '+' should be on a new line. [OperatorWrap]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/config/BTreeIndexConfig.java:132:58: '+' should be on a new line. [OperatorWrap]
--------------------------
</pre>
</details>

hz-devops-test commented:
The job `Hazelcast-pr-compiler` of your PR failed. (Hazelcast internal details: [build log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-compiler/2077/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-compiler/2077/Hazelcast-pr-compiler-2077.zip)).
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
---------SUMMARY----------
--------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-checkstyle-plugin:3.1.2:checkstyle (default) on project hazelcast: An error has occurred in Checkstyle report generation. Failed during checkstyle execution: There are 3 errors reported by Checkstyle 8.38 with /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-compiler/checkstyle/checkstyle.xml ruleset. -> [Help 1]
--------------------------
---------ERRORS-----------
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-compiler/hazelcast/src/main/java/com/hazelcast/config/BTreeIndexConfig.java:130:36: '+' should be on a new line. [OperatorWrap]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-compiler/hazelcast/src/main/java/com/hazelcast/config/BTreeIndexConfig.java:131:40: '+' should be on a new line. [OperatorWrap]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-compiler/hazelcast/src/main/java/com/hazelcast/config/BTreeIndexConfig.java:132:58: '+' should be on a new line. [OperatorWrap]
--------------------------
</pre>
</details>

hz-devops-test commented:
The job `Hazelcast-pr-builder` of your PR failed. (Hazelcast internal details: [build log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/12064/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/12064/Hazelcast-pr-builder-12064.zip)).
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
-------TEST FAILURE-------
--------------------------
[INFO] Results:
[INFO] 
[ERROR] Errors: 
[ERROR]   ClientConfigXmlGeneratorTest.queryCache:680->newConfigViaGenerator:779 ? InvalidConfiguration
[ERROR]   ClientConfigXmlGeneratorTest.queryCache_withEntryEventListenerClass:696->newConfigViaGenerator:779 ? InvalidConfiguration
[ERROR]   DynamicConfigYamlGeneratorTest>AbstractDynamicConfigGeneratorTest.testAttributesConfigWithStoreClass:128->AbstractDynamicConfigGeneratorTest.testMap:917 ? Hazelcast
[ERROR]   DynamicConfigYamlGeneratorTest>AbstractDynamicConfigGeneratorTest.testAttributesConfigWithStoreFactory:156->AbstractDynamicConfigGeneratorTest.testMap:917 ? Hazelcast
[ERROR]   DynamicConfigYamlGeneratorTest>AbstractDynamicConfigGeneratorTest.testAttributesConfigWithStoreFactoryImplementation:170->AbstractDynamicConfigGeneratorTest.testMap:917 ? Hazelcast
[ERROR]   DynamicConfigYamlGeneratorTest>AbstractDynamicConfigGeneratorTest.testAttributesConfigWithStoreImplementation:142->AbstractDynamicConfigGeneratorTest.testMap:917 ? Hazelcast
[INFO] 
[ERROR] Tests run: 50582, Failures: 0, Errors: 6, Skipped: 238
[INFO] 

[ERROR] There are test failures.
</pre>
</details>

hz-devops-test commented:
The job `Hazelcast-pr-builder` of your PR failed. (Hazelcast internal details: [build log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/12075/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/12075/Hazelcast-pr-builder-12075.zip)).
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
---------SUMMARY----------
--------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-checkstyle-plugin:3.1.2:checkstyle (default) on project hazelcast: An error has occurred in Checkstyle report generation. Failed during checkstyle execution: There is 1 error reported by Checkstyle 8.38 with /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/checkstyle/checkstyle.xml ruleset. -> [Help 1]
--------------------------
---------ERRORS-----------
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/config/BTreeIndexConfig.java:121: Line is longer than 130 characters (found 134). [LineLength]
--------------------------
</pre>
</details>

TomaszGaweda commented on hazelcast/src/main/java/com/hazelcast/config/IndexConfig.java:
```diff
@@ -90,6 +92,7 @@ public IndexConfig(IndexConfig other) {
         this.name = other.name;
         this.type = other.type;
         this.bitmapIndexOptions = other.bitmapIndexOptions == null ? null : new BitmapIndexOptions(other.bitmapIndexOptions);
```
imho for consistency bitmapIndexOptions -> bitmapIndexConfig or breeIndexConfig -> btreeIndexOptions. I've chosen BTreeIndexConfig as it's most consistent with other configuration options, but I can change it at any time.

(maybe rename bitmapIndexOptions -> bitmapIndexConfig and make backward-compatibility changes?)

CC @petrpleshachkov @taburet @frant-hartm

hz-devops-test commented:
The job `Hazelcast-pr-compiler` of your PR failed. (Hazelcast internal details: [build log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-compiler/2093/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-compiler/2093/Hazelcast-pr-compiler-2093.zip)).
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
---------SUMMARY----------
--------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-checkstyle-plugin:3.1.2:checkstyle (default) on project hazelcast: An error has occurred in Checkstyle report generation. Failed during checkstyle execution: There is 1 error reported by Checkstyle 8.38 with /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-compiler/checkstyle/checkstyle.xml ruleset. -> [Help 1]
--------------------------
---------ERRORS-----------
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-compiler/hazelcast/src/main/java/com/hazelcast/config/BTreeIndexConfig.java:121: Line is longer than 130 characters (found 134). [LineLength]
--------------------------
</pre>
</details>

hz-devops-test commented:
The job `Hazelcast-pr-builder` of your PR failed. (Hazelcast internal details: [build log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/12076/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/12076/Hazelcast-pr-builder-12076.zip)).
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
-------TEST FAILURE-------
--------------------------
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   DataSerializableConventionsTest.test_identifiedDataSerializables_areInstancesOfSameClass_whenConstructedFromFactory:283 Factory with ID -31 returned null for type with ID 67
[INFO] 
[ERROR] Tests run: 4303, Failures: 1, Errors: 0, Skipped: 8
[INFO] 

[ERROR] There are test failures.
</pre>
</details>

hz-devops-test commented:
The job `Hazelcast-pr-builder` of your PR failed. (Hazelcast internal details: [build log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/12085/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/12085/Hazelcast-pr-builder-12085.zip)).
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
---------SUMMARY----------
--------------------------
[ERROR] COMPILATION ERROR : 
--------------------------
[ERROR] /appdisk/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/internal/config/BTreeIndexConfigReadOnly.java:[32,17] setMemoryTierConfig(com.hazelcast.config.MemoryTierConfig) in com.hazelcast.internal.config.BTreeIndexConfigReadOnly cannot override setMemoryTierConfig(com.hazelcast.config.MemoryTierConfig) in com.hazelcast.config.BTreeIndexConfig
--------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.10.0:compile (default-compile) on project hazelcast: Compilation failure: Compilation failure: 
--------------------------
---------ERRORS-----------
--------------------------
[ERROR] /appdisk/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/internal/config/BTreeIndexConfigReadOnly.java:[32,17] setMemoryTierConfig(com.hazelcast.config.MemoryTierConfig) in com.hazelcast.internal.config.BTreeIndexConfigReadOnly cannot override setMemoryTierConfig(com.hazelcast.config.MemoryTierConfig) in com.hazelcast.config.BTreeIndexConfig
--------------------------
[ERROR] /appdisk/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/internal/config/BTreeIndexConfigReadOnly.java:[31,5] method does not override or implement a method from a supertype
--------------------------
[ERROR] /appdisk/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/internal/config/BTreeIndexConfigReadOnly.java:[37,17] setPageSize(com.hazelcast.memory.Capacity) in com.hazelcast.internal.config.BTreeIndexConfigReadOnly cannot override setPageSize(com.hazelcast.memory.Capacity) in com.hazelcast.config.BTreeIndexConfig
--------------------------
[ERROR] /appdisk/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/internal/config/BTreeIndexConfigReadOnly.java:[36,5] method does not override or implement a method from a supertype
--------------------------
[ERROR] /appdisk/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/internal/config/BTreeIndexConfigReadOnly.java:[32,17] setMemoryTierConfig(com.hazelcast.config.MemoryTierConfig) in com.hazelcast.internal.config.BTreeIndexConfigReadOnly cannot override setMemoryTierConfig(com.hazelcast.config.MemoryTierConfig) in com.hazelcast.config.BTreeIndexConfig
--------------------------
[ERROR] /appdisk/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/internal/config/BTreeIndexConfigReadOnly.java:[31,5] method does not override or implement a method from a supertype
--------------------------
[ERROR] /appdisk/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/internal/config/BTreeIndexConfigReadOnly.java:[37,17] setPageSize(com.hazelcast.memory.Capacity) in com.hazelcast.internal.config.BTreeIndexConfigReadOnly cannot override setPageSize(com.hazelcast.memory.Capacity) in com.hazelcast.config.BTreeIndexConfig
--------------------------
[ERROR] /appdisk/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/internal/config/BTreeIndexConfigReadOnly.java:[36,5] method does not override or implement a method from a supertype
--------------------------
</pre>
</details>

hz-devops-test commented:
The job `Hazelcast-pr-EE-compiler` of your PR failed. (Hazelcast internal details: [build log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-EE-compiler/1486/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-EE-compiler/1486/Hazelcast-pr-EE-compiler-1486.zip)).
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
---------SUMMARY----------
--------------------------
[ERROR] COMPILATION ERROR : 
--------------------------
[ERROR] /appdisk/jenkins/jenkins_slave/workspace/Hazelcast-pr-EE-compiler/hazelcast/hazelcast/src/main/java/com/hazelcast/internal/config/BTreeIndexConfigReadOnly.java:[32,17] setMemoryTierConfig(com.hazelcast.config.MemoryTierConfig) in com.hazelcast.internal.config.BTreeIndexConfigReadOnly cannot override setMemoryTierConfig(com.hazelcast.config.MemoryTierConfig) in com.hazelcast.config.BTreeIndexConfig
--------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.10.0:compile (default-compile) on project hazelcast: Compilation failure: Compilation failure: 
--------------------------
---------ERRORS-----------
--------------------------
[ERROR] /appdisk/jenkins/jenkins_slave/workspace/Hazelcast-pr-EE-compiler/hazelcast/hazelcast/src/main/java/com/hazelcast/internal/config/BTreeIndexConfigReadOnly.java:[32,17] setMemoryTierConfig(com.hazelcast.config.MemoryTierConfig) in com.hazelcast.internal.config.BTreeIndexConfigReadOnly cannot override setMemoryTierConfig(com.hazelcast.config.MemoryTierConfig) in com.hazelcast.config.BTreeIndexConfig
--------------------------
[ERROR] /appdisk/jenkins/jenkins_slave/workspace/Hazelcast-pr-EE-compiler/hazelcast/hazelcast/src/main/java/com/hazelcast/internal/config/BTreeIndexConfigReadOnly.java:[31,5] method does not override or implement a method from a supertype
--------------------------
[ERROR] /appdisk/jenkins/jenkins_slave/workspace/Hazelcast-pr-EE-compiler/hazelcast/hazelcast/src/main/java/com/hazelcast/internal/config/BTreeIndexConfigReadOnly.java:[37,17] setPageSize(com.hazelcast.memory.Capacity) in com.hazelcast.internal.config.BTreeIndexConfigReadOnly cannot override setPageSize(com.hazelcast.memory.Capacity) in com.hazelcast.config.BTreeIndexConfig
--------------------------
[ERROR] /appdisk/jenkins/jenkins_slave/workspace/Hazelcast-pr-EE-compiler/hazelcast/hazelcast/src/main/java/com/hazelcast/internal/config/BTreeIndexConfigReadOnly.java:[36,5] method does not override or implement a method from a supertype
--------------------------
[ERROR] /appdisk/jenkins/jenkins_slave/workspace/Hazelcast-pr-EE-compiler/hazelcast/hazelcast/src/main/java/com/hazelcast/internal/config/BTreeIndexConfigReadOnly.java:[32,17] setMemoryTierConfig(com.hazelcast.config.MemoryTierConfig) in com.hazelcast.internal.config.BTreeIndexConfigReadOnly cannot override setMemoryTierConfig(com.hazelcast.config.MemoryTierConfig) in com.hazelcast.config.BTreeIndexConfig
--------------------------
[ERROR] /appdisk/jenkins/jenkins_slave/workspace/Hazelcast-pr-EE-compiler/hazelcast/hazelcast/src/main/java/com/hazelcast/internal/config/BTreeIndexConfigReadOnly.java:[31,5] method does not override or implement a method from a supertype
--------------------------
[ERROR] /appdisk/jenkins/jenkins_slave/workspace/Hazelcast-pr-EE-compiler/hazelcast/hazelcast/src/main/java/com/hazelcast/internal/config/BTreeIndexConfigReadOnly.java:[37,17] setPageSize(com.hazelcast.memory.Capacity) in com.hazelcast.internal.config.BTreeIndexConfigReadOnly cannot override setPageSize(com.hazelcast.memory.Capacity) in com.hazelcast.config.BTreeIndexConfig
--------------------------
[ERROR] /appdisk/jenkins/jenkins_slave/workspace/Hazelcast-pr-EE-compiler/hazelcast/hazelcast/src/main/java/com/hazelcast/internal/config/BTreeIndexConfigReadOnly.java:[36,5] method does not override or implement a method from a supertype
--------------------------
</pre>
</details>

hz-devops-test commented:
The job `Hazelcast-pr-compiler` of your PR failed. (Hazelcast internal details: [build log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-compiler/2101/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-compiler/2101/Hazelcast-pr-compiler-2101.zip)).
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
---------SUMMARY----------
--------------------------
[ERROR] COMPILATION ERROR : 
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-compiler/hazelcast/src/main/java/com/hazelcast/internal/config/BTreeIndexConfigReadOnly.java:[32,17] setMemoryTierConfig(com.hazelcast.config.MemoryTierConfig) in com.hazelcast.internal.config.BTreeIndexConfigReadOnly cannot override setMemoryTierConfig(com.hazelcast.config.MemoryTierConfig) in com.hazelcast.config.BTreeIndexConfig
--------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.10.0:compile (default-compile) on project hazelcast: Compilation failure: Compilation failure: 
--------------------------
---------ERRORS-----------
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-compiler/hazelcast/src/main/java/com/hazelcast/internal/config/BTreeIndexConfigReadOnly.java:[32,17] setMemoryTierConfig(com.hazelcast.config.MemoryTierConfig) in com.hazelcast.internal.config.BTreeIndexConfigReadOnly cannot override setMemoryTierConfig(com.hazelcast.config.MemoryTierConfig) in com.hazelcast.config.BTreeIndexConfig
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-compiler/hazelcast/src/main/java/com/hazelcast/internal/config/BTreeIndexConfigReadOnly.java:[31,5] method does not override or implement a method from a supertype
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-compiler/hazelcast/src/main/java/com/hazelcast/internal/config/BTreeIndexConfigReadOnly.java:[37,17] setPageSize(com.hazelcast.memory.Capacity) in com.hazelcast.internal.config.BTreeIndexConfigReadOnly cannot override setPageSize(com.hazelcast.memory.Capacity) in com.hazelcast.config.BTreeIndexConfig
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-compiler/hazelcast/src/main/java/com/hazelcast/internal/config/BTreeIndexConfigReadOnly.java:[36,5] method does not override or implement a method from a supertype
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-compiler/hazelcast/src/main/java/com/hazelcast/internal/config/BTreeIndexConfigReadOnly.java:[32,17] setMemoryTierConfig(com.hazelcast.config.MemoryTierConfig) in com.hazelcast.internal.config.BTreeIndexConfigReadOnly cannot override setMemoryTierConfig(com.hazelcast.config.MemoryTierConfig) in com.hazelcast.config.BTreeIndexConfig
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-compiler/hazelcast/src/main/java/com/hazelcast/internal/config/BTreeIndexConfigReadOnly.java:[31,5] method does not override or implement a method from a supertype
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-compiler/hazelcast/src/main/java/com/hazelcast/internal/config/BTreeIndexConfigReadOnly.java:[37,17] setPageSize(com.hazelcast.memory.Capacity) in com.hazelcast.internal.config.BTreeIndexConfigReadOnly cannot override setPageSize(com.hazelcast.memory.Capacity) in com.hazelcast.config.BTreeIndexConfig
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-compiler/hazelcast/src/main/java/com/hazelcast/internal/config/BTreeIndexConfigReadOnly.java:[36,5] method does not override or implement a method from a supertype
--------------------------
</pre>
</details>

TomaszGaweda commented:
run-lab-run

taburet commented on hazelcast/src/main/java/com/hazelcast/config/BTreeIndexConfig.java:
```diff
@@ -0,0 +1,139 @@
+    /**
+     * Default value of page size for B-Trees.
+     */
+    public static final Capacity DEFAULT_PAGE_SIZE = Capacity.of(1024, MemoryUnit.BYTES);
```
I think for t-store b+tree we should raise the default to something like 8-64kB at least, let's start with 16, performance testing will tell us a proper default later.

For the regular HD b+tree `GlobalIndexPoolingAllocator#DEFAULT_BTREE_INDEX_NODE_SIZE` is used as a default currently, which is 8kB. Would the new setting deprecate/override it? If yes, we probably should have different defaults for the regular and t-store b+trees.

BTW, looks like a natural unit here is kilobytes.

taburet commented:
@blazember wdyt are there any benefits in explicit configuration option for hybrid log page size? are there any benefits at all in making it bigger than b+tree page?

petrpleshachkov commented on hazelcast/src/main/java/com/hazelcast/client/impl/protocol/codec/custom/BTreeIndexConfigCodec.java:
```diff
@@ -0,0 +1,53 @@
+import static com.hazelcast.client.impl.protocol.ClientMessage.*;
+import static com.hazelcast.client.impl.protocol.codec.builtin.FixedSizeTypesCodec.*;
+
+@Generated("e821dbe123d55d5e1f205014b5e02135")
```
Not sure, do we add generated code to the codebase?

TomaszGaweda replied:
Yes, each time the protocol changes you have to run Python script to regenerate scripts

petrpleshachkov reviewed (commented):
Do we do anywhere a validation that a tiered index created on a non-tiered map is not allowed?

petrpleshachkov reviewed (approved):
Approve with the validation comment

TomaszGaweda commented:
A.d. validation: it would be nice to have, but by-convention our getSthConfig() always return empty object. Therefore, it will return something even if user didn't touch it. I can duplicate the information is it was enabled manually (e.g. setEnabled()), but IMHO it's not that user friendly neither. Better to be silent if map is not tiered and there is some config, IIRC not the first place where this approach is used.
Discussed on DMs

mdumandag commented:
@TomaszGaweda Did you go over this checklist and made sure that everything is there? https://github.com/hazelcast/hazelcast/wiki/Checklist-for-Hazelcast-config-changes

If it is the case, I can start reviewing the PR

TomaszGaweda commented:
It should be all ok @mdumandag, I've added Spring configuration only, because I wrongly assumed that it's not-Spring specific, so I don't need to check that :D But I fixed that.

hz-devops-test commented:
The job `Hazelcast-pr-builder` of your PR failed. (Hazelcast internal details: [build log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/12332/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/12332/Hazelcast-pr-builder-12332.zip)).
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
-------TEST FAILURE-------
--------------------------
[INFO] Results:
[INFO] 
[ERROR] Errors: 
[ERROR]   TestFullApplicationContext.testAuditlogConfig ? IllegalState Failed to load Ap...
[ERROR]   TestFullApplicationContext.testCPSubsystemConfig ? IllegalState Failed to load...
[ERROR]   TestFullApplicationContext.testCRDTReplicationConfig ? IllegalState Failed to ...
[ERROR]   TestFullApplicationContext.testCacheConfig ? IllegalState Failed to load Appli...
[ERROR]   TestFullApplicationContext.testCardinalityEstimatorConfig ? IllegalState Faile...
[ERROR]   TestFullApplicationContext.testClusterNameConfig ? IllegalState Failed to load...
[ERROR]   TestFullApplicationContext.testCompactSerializationConfig ? IllegalState Faile...
[ERROR]   TestFullApplicationContext.testConfigListeners ? IllegalState Failed to load A...
[ERROR]   TestFullApplicationContext.testDevice ? IllegalState Failed to load Applicatio...
[ERROR]   TestFullApplicationContext.testDurableExecutorConfig ? IllegalState Failed to ...
[ERROR]   TestFullApplicationContext.testDynamicConfiguration ? IllegalState Failed to l...
[ERROR]   TestFullApplicationContext.testExecutorConfig ? IllegalState Failed to load Ap...
[ERROR]   TestFullApplicationContext.testExplicitPortCountConfiguration ? IllegalState F...
[ERROR]   TestFullApplicationContext.testFullQueryCacheConfig ? IllegalState Failed to l...
[ERROR]   TestFullApplicationContext.testHazelcastInstances ? IllegalState Failed to loa...
[ERROR]   TestFullApplicationContext.testInstance ? IllegalState Failed to load Applicat...
[ERROR]   TestFullApplicationContext.testInstanceTrackingConfig ? IllegalState Failed to...
[ERROR]   TestFullApplicationContext.testIntegrityCheckerConfig ? IllegalState Failed to...
[ERROR]   TestFullApplicationContext.testJavaSerializationFilterConfig ? IllegalState Fa...
[ERROR]   TestFullApplicationContext.testJetConfig ? IllegalState Failed to load Applica...
[ERROR]   TestFullApplicationContext.testListConfig ? IllegalState Failed to load Applic...
[ERROR]   TestFullApplicationContext.testManagementCenterConfig ? IllegalState Failed to...
[ERROR]   TestFullApplicationContext.testMapConfig ? IllegalState Failed to load Applica...
[ERROR]   TestFullApplicationContext.testMapEvictionPolicies ? IllegalState Failed to lo...
[ERROR]   TestFullApplicationContext.testMapEvictionPolicyClassName ? IllegalState Faile...
[ERROR]   TestFullApplicationContext.testMapEvictionPolicyImpl ? IllegalState Failed to ...
[ERROR]   TestFullApplicationContext.testMapNativeMaxSizePolicy ? IllegalState Failed to...
[ERROR]   TestFullApplicationContext.testMapNoWanMergePolicy ? IllegalState Failed to lo...
[ERROR]   TestFullApplicationContext.testMapPerEntryStats ? IllegalState Failed to load ...
[ERROR]   TestFullApplicationContext.testMemberAttributesConfig ? IllegalState Failed to...
[ERROR]   TestFullApplicationContext.testMemberFlakeIdGeneratorConfig ? IllegalState Fai...
[ERROR]   TestFullApplicationContext.testMemberNearCacheEvictionPolicies ? IllegalState ...
[ERROR]   TestFullApplicationContext.testMemcacheProtocolConfig ? IllegalState Failed to...
[ERROR]   TestFullApplicationContext.testMetricsConfig ? IllegalState Failed to load App...
[ERROR]   TestFullApplicationContext.testMultimapConfig ? IllegalState Failed to load Ap...
[ERROR]   TestFullApplicationContext.testNativeMemoryConfig ? IllegalState Failed to loa...
[ERROR]   TestFullApplicationContext.testNetworkConfig ? IllegalState Failed to load App...
[ERROR]   TestFullApplicationContext.testNoWanConsumerClass ? IllegalState Failed to loa...
[ERROR]   TestFullApplicationContext.testPNCounterConfig ? IllegalState Failed to load A...
[ERROR]   TestFullApplicationContext.testPartitionGroupConfig ? IllegalState Failed to l...
[ERROR]   TestFullApplicationContext.testPersistence ? IllegalState Failed to load Appli...
[ERROR]   TestFullApplicationContext.testProbabilisticSplitBrainProtectionConfig ? IllegalState
[ERROR]   TestFullApplicationContext.testProperties ? IllegalState Failed to load Applic...
[ERROR]   TestFullApplicationContext.testQueueConfig ? IllegalState Failed to load Appli...
[ERROR]   TestFullApplicationContext.testRecentlyActiveSplitBrainProtectionConfig ? IllegalState
[ERROR]   TestFullApplicationContext.testReliableTopicConfig ? IllegalState Failed to lo...
[ERROR]   TestFullApplicationContext.testReplicatedMapConfig ? IllegalState Failed to lo...
[ERROR]   TestFullApplicationContext.testRestApiConfig ? IllegalState Failed to load App...
[ERROR]   TestFullApplicationContext.testRingbufferConfig ? IllegalState Failed to load ...
[ERROR]   TestFullApplicationContext.testSSLConfig ? IllegalState Failed to load Applica...
[ERROR]   TestFullApplicationContext.testScheduledExecutorConfig ? IllegalState Failed t...
[ERROR]   TestFullApplicationContext.testSecurity ? IllegalState Failed to load Applicat...
[ERROR]   TestFullApplicationContext.testSerializationConfig ? IllegalState Failed to lo...
[ERROR]   TestFullApplicationContext.testSetConfig ? IllegalState Failed to load Applica...
[ERROR]   TestFullApplicationContext.testSocketInterceptorConfig ? IllegalState Failed t...
[ERROR]   TestFullApplicationContext.testSplitBrainProtectionConfig ? IllegalState Faile...
[ERROR]   TestFullApplicationContext.testSqlConfig ? IllegalState Failed to load Applica...
[ERROR]   TestFullApplicationContext.testTopicConfig ? IllegalState Failed to load Appli...
[ERROR]   TestFullApplicationContext.testWanConsumerWithPersistDataFalse ? IllegalState ...
[ERROR]   TestFullApplicationContext.testWanReplicationConfig ? IllegalState Failed to l...
[ERROR]   TestFullApplicationContext.testWanReplicationSyncConfig ? IllegalState Failed ...
[ERROR]   TestFullApplicationContext.testWhenBothMapEvictionPolicyClassNameAndEvictionPolicySet ? IllegalState
[INFO] 
[ERROR] Tests run: 207, Failures: 0, Errors: 62, Skipped: 0
[INFO] 

[ERROR] There are test failures.
</pre>
</details>

TomaszGaweda commented:
Previous test failures was just an oversight, next job will be green :)

mdumandag commented on hazelcast/src/main/java/com/hazelcast/internal/dynamicconfig/DynamicConfigYamlGenerator.java:
```diff
@@ -965,6 +966,15 @@ private static List<Map<String, Object>> getIndexConfigsAsList(List<IndexConfig>
+                Map<String, Object> capacityAsMap = new LinkedHashMap<>();
+                addNonNullToMap(capacityAsMap, "capacity", getCapacityAsMap(bTreeConf.getMemoryTierConfig().getCapacity()));
+
+                addNonNullToMap(btreeOptionsAsMap, "memory-tier", capacityAsMap);
```
What do we this btreeOptionsAsMap? I see we have populated it, but didn't put it to anywhere

mdumandag replied:
Can we add a test for this to AbstractDynamicConfigGeneratorTest?

TomaszGaweda replied:
Added, thx

mdumandag commented on hazelcast/src/main/java/com/hazelcast/internal/config/BTreeIndexConfigReadOnly.java:
```diff
@@ -0,0 +1,40 @@
+/**
+ * A read-only counterpart of {@link BTreeIndexConfig} class.
+ */
+public class BTreeIndexConfigReadOnly extends BTreeIndexConfig {
```
Where do we use this class?

mdumandag commented on hazelcast-spring/src/main/resources/hazelcast-spring-5.2.xsd:
```diff
@@ -3830,6 +3831,13 @@
 
+    <xs:complexType name="btree-index">
+        <xs:all>
+            <xs:element name="page-size" type="memory-size" minOccurs="1"/>
```
Why does `minOccurs="1"`? Why don't we let the user omit this value if they want to use the default value for it? Also, can you add the documentation in hazelcast-config-5.2.xsd to here as well?

mdumandag commented on hazelcast/src/main/resources/hazelcast-config-5.2.xsd:
```diff
@@ -2534,10 +2534,11 @@
     </xs:complexType>
     <xs:complexType name="index">
-        <xs:all>
+        <xs:sequence>
```
Isn't this breaking change? Before that something like 
```
<bitmap-index-options>....
<attributes>...
```
was legal, but now we enforce order of tags with xs:sequence

TomaszGaweda replied:
True, reverted

mdumandag commented on hazelcast/src/main/resources/hazelcast-full-example.xml:
```diff
@@ -1404,6 +1404,17 @@
+                    </memory-tier>
+                    <page-size value="1024" unit="BYTES" />
+                </btree-index>
+            </index>
```
Not related to this, but can you add a similar sample to hazelcast-fullconfig-without-network.[xml&yaml]

mdumandag commented on hazelcast-spring-tests/src/test/java/com/hazelcast/spring/TestFullApplicationContext.java:
```diff
@@ -371,15 +373,20 @@ public void testMapConfig() {
+            } else if ("sortedIndex".equals(index.getName())) {
                 assertEquals(IndexType.SORTED, index.getType());
-                assertEquals("sortedIndex", index.getName());
-                assertEquals("name", index.getAttributes().get(1));
```
minor: can we add `assertEquals("name", index.getAttributes().get(1));` assertion back?

mdumandag commented on hazelcast-spring/src/main/java/com/hazelcast/spring/AbstractHazelcastBeanDefinitionParser.java:
```diff
@@ -655,14 +658,37 @@ protected void handleIndex(ManagedList<BeanDefinition> indexes, Node indexNode)
+                        if ("page-size".equals(cleanNodeName(columnNode))) {
+                            treeIndexConfig.setPageSize(parseCapacity(columnNode));
+                        }
+                        if ("memory-tier".equals(cleanNodeName(columnNode))) {
```
minor: could be `else if`

mdumandag commented on hazelcast/src/main/java/com/hazelcast/config/BTreeIndexConfig.java:
```diff
@@ -0,0 +1,140 @@
+
+    private Capacity pageSize = DEFAULT_PAGE_SIZE;
+
+    private MemoryTierConfig memoryTierConfig;
```
Is `null` a valid value for this field? We made it non-nullable in the protocol repo

TomaszGaweda replied:
when you call getMemoryTierConfig() it will fill this field if it was null

mdumandag commented on hazelcast/src/main/java/com/hazelcast/config/BTreeIndexConfig.java:
```diff
@@ -0,0 +1,140 @@
+     * @param pageSize page size of the index
+     */
+    public BTreeIndexConfig setPageSize(Capacity pageSize) {
+        this.pageSize = pageSize;
```
Should we have a null check here? We made this field non-nullable in the protocol repo

mdumandag commented on hazelcast/src/main/java/com/hazelcast/config/BTreeIndexConfig.java:
```diff
@@ -0,0 +1,140 @@
+     * @param memoryTierConfig new memory tier configuration to be set.
+     */
+    public BTreeIndexConfig setMemoryTierConfig(MemoryTierConfig memoryTierConfig) {
+        this.memoryTierConfig = memoryTierConfig;
```
Should we have a null check here? We made this field non-nullable in the protocol repo

mdumandag commented on hazelcast/src/main/java/com/hazelcast/config/IndexConfig.java:
```diff
@@ -218,6 +221,27 @@ public IndexConfig setBitmapIndexOptions(BitmapIndexOptions bitmapIndexOptions)
+     * @return the b-tree index configuration associated with this config.
+     */
+    public BTreeIndexConfig getBTreeIndexConfig() {
+        if (bTreeIndexConfig == null) {
```
I didn't get why this field can be null. Why don't we use the `new BTreeIndexConfig()` as the default value and not allow user to set it to null?

mdumandag replied:
I see. I couldn't think of a scenario where having this null by default would help other than maybe trying to carry fewer bytes over the network when this index config is serialized unless the user sets it to something non-default.

TomaszGaweda replied:
We've discussed this on daily, there is no hard convention, so I will follow your comment - it's simpler :)

hz-devops-test commented:
The job `Hazelcast-pr-builder` of your PR failed. (Hazelcast internal details: [build log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/12336/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/12336/Hazelcast-pr-builder-12336.zip)).
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
-------TEST FAILURE-------
--------------------------
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   TestFullApplicationContext.testMapConfig:388 expected:<129 GIGABYTES> but was:<256 MEGABYTES>
[INFO] 
[ERROR] Tests run: 207, Failures: 1, Errors: 0, Skipped: 0
[INFO] 

[ERROR] There are test failures.
</pre>
</details>

mdumandag commented on hazelcast/src/main/java/com/hazelcast/query/impl/IndexUtils.java:
```diff
@@ -333,11 +351,31 @@ public static IndexConfig getIndexConfigFromXml(Node indexNode, boolean domLevel
                 res.getBitmapIndexOptions().setUniqueKeyTransformation(uniqueKeyTransformation);
             }
+        } else if (type == IndexType.SORTED) {
+            Node optionsNode = childElementWithName(indexNode, "btree-index", strict);
```
I see that if a user uses this new config with other index types, we simply ignore it. Would it be better to check this and fail-fast for invalid configurations? I am not sure which behavior is better, so don't hesitate to ignore this comment if you think the current behavior is better

TomaszGaweda replied:
IIRC we always ignore config that is not relevant

mdumandag reviewed (approved):
Approving with a few minor comments left. I left them unresolved, but as I said, they are really minor and we can simply ignore them
